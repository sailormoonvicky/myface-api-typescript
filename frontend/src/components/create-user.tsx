import { useState, FormEventHandler, ChangeEventHandler } from "react";
import { CreateUserRequest } from "../../../src/models/api/createUserRequest";

export const CreateUser:React.FC = () => {
    const [inputs, setInputs] = useState<CreateUserRequest>({
        name: undefined || '',
        username: undefined || '',
        profileImageUrl: undefined || '',
        coverImageUrl: undefined || '',
        email: undefined || ''
    });

    const handleChange: ChangeEventHandler<HTMLInputElement> = (event) => {
        const {name, value} = event.target;
        setInputs(values => ({...values, [name]: value}))
      
    }
    const usernameCheckValidity =() => {
        const regex = /^[a-z]+$/;
        return regex.test(inputs.username);
      }
    
      const emailCheckValidity =() => {
        const regex = /[.'_%\-+A-Za-z0-9]+@\w+([-.]?\w+)*(\.\w{2,3})+/;
        return regex.test(inputs.email);
      }
    
    const validateForm = () => {
        let errMsg = '';
        let ifValid = true;
        //Check none of the values are null
        Object.entries(inputs).forEach(([k, v]) => {
            if(v===undefined || v===''){
                errMsg = errMsg + 'Please enter ' + k + '<br />';
                ifValid = false;      
            }
        });

        //Check for valid email id
        if(!emailCheckValidity()){
            ifValid = false;
            errMsg = errMsg + 'Please enter a valid Email ID' + '<br />';
        }
        //Check for valid email id
        if(!usernameCheckValidity()){
            ifValid = false;
            errMsg = errMsg + 'Please enter a valid username (lower case letters only without spaces)' + '<br />';
        }
        console.log('-->',errMsg)
        let element = document.getElementById("divError");
        if(element!==null)
            {element.innerHTML = errMsg;}
        return ifValid;        
      };

    const handleSubmit:FormEventHandler<HTMLFormElement> = (e) => {
        e.preventDefault();
        
        if(validateForm()){
            fetch("http://localhost:3001/users/create/", {
            method: 'POST',
            body: JSON.stringify(inputs),   
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },          
            })
            .then((response) => {
                if (response.status !== 200) {
                throw new Error(response.statusText);
                }                   
            })
            .then(() => {
                console.log("We'll be in touch soon.");
            })
            .catch((err) => {
                console.log(err.toString());
            });
        }
      };

    return (
        <>
        <form onSubmit={handleSubmit}>
            <label>Enter your name:
            <input 
            type="text" 
            name="name" 
            value={inputs.name || ""} 
            onChange={handleChange}
            />
            </label>
            <label>Enter your username:
            <input 
                type="text" 
                name="username" 
                value={inputs.username || ""} 
                onChange={handleChange}
            />
            </label>
            <label>Enter your Profile Image URL:
            <input 
                type="text" 
                name="profileImageUrl" 
                value={inputs.profileImageUrl || ""} 
                onChange={handleChange}
            />
            </label>
            <label>Enter your Cover Image URL:
            <input 
                type="text" 
                name="coverImageUrl" 
                value={inputs.coverImageUrl || ""} 
                onChange={handleChange}
            />
            </label>
            <label>Enter your Email ID:
            <input 
                type="email" 
                name="email" 
                value={inputs.email || ""} 
                onChange={handleChange}
            />
            </label>
          <button type="submit">Submit</button>
        </form>
        <div id="divError"></div>
        </>
    )
}

