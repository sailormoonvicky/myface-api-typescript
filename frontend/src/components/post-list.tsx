import React, {useEffect, useState} from "react";


export const PostGallery: React.FC = () => {
    const [postLists, setPostLists] = useState<any[]>([]);

    useEffect(() => {
        fetch('http://localhost:3001/posts')
        .then(response => response.json())
        .then(data =>setPostLists(data.results))
    }, [])


    return (
        <>
            <div className="banner"><h1>Posts</h1></div>
            <div className="posts">
                <ol>
                {postLists.map((post) => (
                    <li key={post.id}>
                        <img src={post.imageUrl} alt={`A image with post id ${post.id}`} />
                        <div>
                            <p className="username">by {post.postedBy.username}</p>
                            <p className="timestamp">{post.createdAt}</p>
                            <p className="message">{post.message}</p>
                        </div>
                    </li>
                ))}
                </ol>
            </div>
        </>
    )
}