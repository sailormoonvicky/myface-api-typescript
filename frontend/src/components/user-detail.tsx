import React, {useEffect, useState} from "react";


export const UserDetail: React.FC<{userID: number}> = (props:{userID: number}) => {
    const [userDetail, setuserDetail] = useState<any>({});

    useEffect(() => {
        fetch(`http://localhost:3001/users/${props.userID}`)
        .then(response => response.json())
        .then(data => setuserDetail(data))
    }, [props.userID])

    return(
        <>
        <ul>
            <li>ID: {userDetail.id}</li>
            <li>Username: {userDetail.username}</li>
            <li>Full Name: {userDetail.name}</li>
            <li>Email: {userDetail.email}</li>
            <li><img src={userDetail.coverImageUrl} alt="This is profile image"/></li>
            <li><img src={userDetail.profileImageUrl} alt="This is cover image"/></li>
        </ul>
        </>
    )
}