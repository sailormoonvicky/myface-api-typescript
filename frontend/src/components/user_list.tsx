import React, {useEffect, useState} from "react";
import {Link} from 'react-router-dom';

export const UserList: React.FC<{setuserID: React.Dispatch<React.SetStateAction<number>>}> 
                                = (props:{setuserID: React.Dispatch<React.SetStateAction<number>>}) => {
    const [userLists, setUserLists] = useState<any[]>([]);

    useEffect(() => {
        fetch('http://localhost:3001/users')
        .then(response => response.json())
        .then(data =>setUserLists(data.results))
    }, [])

    return (
        <>
            <h1>Users</h1>
            <ol>
                {userLists.map(user => (
                    <li key={user.id}>
                        <Link to={`/users/${user.id}`} onClick={() => props.setuserID(user.id)}>
                        {user.name}
                        </Link>
                        <p>{user.id}</p>
                        <p>{user.username}</p>
                        <img src={user.profileImageUrl} alt="An user profile image" />
                        <p>Post: {user.posts.length} Likes: {user.likes.length} Dislikes: {user.dislikes.length} </p>
                    </li>
                ))}
            </ol>
        </>
    )

}