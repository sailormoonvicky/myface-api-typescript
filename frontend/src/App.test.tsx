
import { render } from '@testing-library/react';
// import App from './App';
import * as renderer from 'react-test-renderer';
import { CreateUser } from './components/create-user';
import { PostGallery} from './components/post-list';
import { UserDetail } from './components/user-detail';
//import { UserList } from './components/user_list';

describe('Regression / Snapshot test', function () {    
    test('Snapshot for Create User', async function () {
        const tree = renderer.create(<CreateUser />).toJSON();
        expect(tree).toMatchSnapshot();   
    });
    test('Snapshot for User Detail', async function () {
        const tree = renderer.create(<UserDetail userID={1}/>).toJSON();
        expect(tree).toMatchSnapshot();   
    });
    test('Snapshot for Post List', async function () {
        const tree = renderer.create(<PostGallery />).toJSON();
        expect(tree).toMatchSnapshot();   
    });
    // test('Snapshot for User List', async function () {
    //     const [, setuserID] = useState(0);
    //     const tree = renderer.create(<UserList setuserID = {setuserID}></UserList>).toJSON();
    //     expect(tree).toMatchSnapshot();   
    // });
    
});

describe('Component Testing', function () {  
    test('renders Create User', () => {
        const { getByText } = render(<CreateUser />);
        const textElement = getByText(/Enter your name/i);
        expect(textElement).toBeInTheDocument();
    });
    test('renders Post List', () => {
        const { getByText } = render(<PostGallery />);
        const textElement = getByText(/Posts/i);
        expect(textElement).toBeInTheDocument();
    });    
});
