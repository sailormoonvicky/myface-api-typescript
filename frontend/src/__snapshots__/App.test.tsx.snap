// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Regression / Snapshot test Snapshot for App 1`] = `
<h1>
  MyFace
</h1>
`;

exports[`Regression / Snapshot test Snapshot for Create User 1`] = `
[
  <form
    onSubmit={[Function]}
  >
    <label>
      Enter your name:
      <input
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <label>
      Enter your username:
      <input
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <label>
      Enter your Profile Image URL:
      <input
        name="profileImageUrl"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <label>
      Enter your Cover Image URL:
      <input
        name="coverImageUrl"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <label>
      Enter your Email ID:
      <input
        name="email"
        onChange={[Function]}
        type="email"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>,
  <div
    id="divError"
  />,
]
`;

exports[`Regression / Snapshot test Snapshot for Post List 1`] = `
[
  <div
    className="banner"
  >
    <h1>
      Posts
    </h1>
  </div>,
  <div
    className="posts"
  >
    <ol />
  </div>,
]
`;

exports[`Regression / Snapshot test Snapshot for User Detail 1`] = `
<div
  className="flexRow"
>
  <div
    className="flexColumn"
  >
    <img
      alt="This is cover image"
      className="coverImage"
    />
    <div
      className="flexAvatarContainer"
    >
      <img
        alt="This is cover image"
        className="profileImage"
      />
      <div
        className="textContainer"
      >
        <p
          className="fullName"
        />
        <div
          className="alignTextLeft"
        >
          <div
            className="userName"
          />
          <div
            className="email"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="flexColumn"
  >
    <p
      className="heading"
    >
      's Posts
    </p>
    <div
      className="flexCardsContainer"
    />
    <div>
      <button
        onClick={[Function]}
        type="button"
      >
        Load More
      </button>
    </div>
  </div>
  <div
    className="flexColumn"
  >
    <img
      alt="This is cover image"
      className="coverImage"
    />
  </div>
  <div
    className="flexColumn"
  >
    <img
      alt="This is cover image"
      className="coverImage"
    />
  </div>
</div>
`;
